src/sparse/impl/KokkosSparse_spmv_impl.hpp

```diff
@@ -502,380 +502,449 @@ struct SPMV_MV_Transpose_Functor {
-        return;
+    KOKKOS_INLINE_FUNCTION
+    value_type get(const ordinal_type vecIdx) {
+      switch (vecIdx) {
+      case 0:
+        return x0;
+      case 1:
+        return x1;
       }
 
-      // mfh 20 Mar 2015, 07 Jun 2016: This is ordinal_type because it
-      // needs to have the same type as n.
-      ordinal_type kk = 0;
+      return 0;
+    }
+  };
 
-#ifdef KOKKOS_FAST_COMPILE
-      for (; kk + 4 <= n; kk += 4) {
-        strip_mine<4>(dev, iRow, kk);
-      }
-      for( ; kk < n; ++kk) {
-        strip_mine<1>(dev, iRow, kk);
-      }
-#else
-#  ifdef __CUDA_ARCH__
-      if ((n > 8) && (n % 8 == 1)) {
-        strip_mine<9>(dev, iRow, kk);
-        kk += 9;
-      }
-      for(; kk + 8 <= n; kk += 8)
-        strip_mine<8>(dev, iRow, kk);
-      if(kk < n)
-        switch(n - kk) {
-#  else // NOT a CUDA device
-          if ((n > 16) && (n % 16 == 1)) {
-            strip_mine<17>(dev, iRow, kk);
-            kk += 17;
-          }
+  SPMV_MV_Functor (const value_type alpha_,
+                   const AMatrix m_A_,
+                   const XVector m_x_,
+                   const value_type beta_,
+                   const YVector m_y_,
+                   const ordinal_type rows_per_team_) :
+    alpha (alpha_), m_A (m_A_), m_x (m_x_),
+    beta (beta_), m_y (m_y_),
+    rows_per_team (rows_per_team_),
+    numVecs (m_x_.extent(1)),
+    vecRemainder (m_x_.extent(1) % 16)
+  {
+    static_assert (static_cast<int> (XVector::rank) == 2,
+                   "XVector must be a rank 2 View.");
+    static_assert (static_cast<int> (YVector::rank) == 2,
+                   "YVector must be a rank 2 View.");
+  }
 
-          for (; kk + 16 <= n; kk += 16) {
-            strip_mine<16>(dev, iRow, kk);
-          }
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const team_member& dev) const
+  {
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(dev,0,rows_per_team), [&] (const ordinal_type& loop) {
+        const ordinal_type iRow = static_cast<ordinal_type>(dev.league_rank()) * rows_per_team + loop;
+        if (iRow >= m_A.numRows ()) {
+          return;
+        }
+        const KokkosSparse::SparseRowViewConst<AMatrix> row = m_A.rowConst(iRow);
+        const ordinal_type row_length = static_cast<ordinal_type> (row.length);
+        ordinal_type remainder = vecRemainder;
+
+	for(ordinal_type vecOffset = 0; vecOffset < numVecs - 15; vecOffset += 16) {
+	  value16 sum = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
+                         0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
+	  Kokkos::parallel_reduce(Kokkos::ThreadVectorRange(dev, row_length),
+				  [&] (const ordinal_type& entryIdx, value16& lsum) {
+
+				    // Extract data from current row
+				    const value_type val = alpha*(conjugate ? ATV::conj(row.value(entryIdx)) : row.value(entryIdx));
+				    const ordinal_type colIdx = row.colidx(entryIdx);
+
+				    // Perform dot product accros lhs columns
+				    lsum.x0  += m_x(colIdx, vecOffset +  0)*val;
```

Comment:
This feels like it should be a method on `value16` etc., that takes `lsum`, `m_x`, `colIdx`, and `val`.